File: music/data/raw/spotify/spotify-models/src/main/java/yurykorzun/art/universe/music/data/raw/spotify/domain/entity/SpotifyAlbum.java
package yurykorzun.art.universe.music.data.raw.spotify.domain.entity;

import jakarta.persistence.*;
import lombok.*;
import lombok.experimental.SuperBuilder;
import yurykorzun.art.universe.music.data.raw.spotify.common.SpotifyConstants;
import yurykorzun.art.universe.music.data.raw.spotify.domain.entity.common.BaseSpotifyEntity;
import yurykorzun.art.universe.music.data.raw.spotify.enums.SpotifyAlbumType;
import yurykorzun.art.universe.music.data.raw.spotify.enums.SpotifyAlbumTypeConverter;
import yurykorzun.art.universe.music.data.raw.spotify.enums.SpotifyDatePrecision;
import yurykorzun.art.universe.music.data.raw.spotify.enums.SpotifyDatePrecisionConverter;
import yurykorzun.art.universe.music.data.raw.spotify.enums.SpotifyEntityType;

@Entity(name = "album")
@SuperBuilder
@NoArgsConstructor
@Getter @Setter
public class SpotifyAlbum extends BaseSpotifyEntity {

    @Id
    @SequenceGenerator(name = "album_seq_gen", sequenceName = "album_seq", allocationSize = SpotifyConstants.HIBERNATE_BATCH_SIZE)
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "album_seq_gen")
    @Setter(AccessLevel.NONE)
    private long id;

    @Column(name = "album_type")
    @Convert(converter = SpotifyAlbumTypeConverter.class)
    private SpotifyAlbumType albumType;

    @Column(name = "total_tracks")
    private Integer totalTracks;

    @Column(name = "release_date")
    private String releaseDate;

    @Column(name = "release_date_precision")
    @Convert(converter = SpotifyDatePrecisionConverter.class)
    private SpotifyDatePrecision releaseDatePrecision;

    @Column(name = "spotify_url")
    private String spotifyUrl;

    @Column(name = "uri")
    private String uri;

    @Column(name = "primary_artist_id")
    private Long primaryArtistId;

    @Column(name = "primary_artist_spotify_id")
    private String primaryArtistSpotifyId;

    @Override
    public SpotifyEntityType getEntityType() {
        return SpotifyEntityType.ALBUM;
    }
}

File: ui/src/music/data/raw/spotify/types/spotify-album.ts
import { BaseSpotifyEntity } from "./spotify-base-entity.ts";
import { type ArtistRelatedRawEntity } from "@/music/shared/types/entities.ts";
import type { Album } from "@/music/shared/types/entities.ts";

export class SpotifyAlbum
    extends BaseSpotifyEntity<"album">
    implements ArtistRelatedRawEntity<"album">
{
    spotifyId: string;
    spotifyUrl: string | null;
    totalTracks: number | null;
    releaseDate: string | null;
    primaryArtistId: number | null;
    primaryArtistName: string | null;

    constructor(
        id: number,
        name: string,
        spotifyId: string,
        spotifyUrl: string | null,
        totalTracks: number | null,
        releaseDate: string | null,
        primaryArtistId: number | null,
        primaryArtistName: string | null,
        masterEntity?: Album
    ) {
        super(id, name, masterEntity);
        this.spotifyId = spotifyId;
        this.spotifyUrl = spotifyUrl;
        this.totalTracks = totalTracks;
        this.releaseDate = releaseDate;
        this.primaryArtistId = primaryArtistId;
        this.primaryArtistName = primaryArtistName;
    }

    getEntityType(): "album" {
        return "album";
    }

    get artist(): { id: number; name: string } | undefined {
        if (this.primaryArtistId && this.primaryArtistName) {
            return { id: this.primaryArtistId, name: this.primaryArtistName };
        }
        return undefined;
    }

    getExternalArtistId(): number | undefined {
        return this.primaryArtistId ?? undefined;
    }

    getMasterArtistId(): number | undefined {
        return this.getMasterEntity()?.primaryArtistId;
    }
}
